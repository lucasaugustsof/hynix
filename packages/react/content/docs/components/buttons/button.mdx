import { Meta } from '@storybook/blocks'

import buttonStories from '../../../../__stories__/components/buttons/button.stories'

<Meta
  title='components/buttons/Button'
  of={buttonStories}
/>

# Button
