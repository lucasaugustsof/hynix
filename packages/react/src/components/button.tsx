import { type Assign, ark } from '@ark-ui/react'

import { cn } from '@r/utilities/cn'
import { type VariantProps, tv } from '@r/utilities/tv'

type ButtonProps = Assign<
  React.CustomComponentPropsWithRef<typeof ark.button>,
  VariantProps<typeof buttonVariants>
>

const buttonVariants = tv({
  base: [
    [
      'isolate inline-flex cursor-pointer select-none items-center justify-center whitespace-nowrap rounded-xl font-sans font-semibold tracking-normal',
      'not-disabled:inset-ring-1 disabled:cursor-not-allowed disabled:bg-fill-1 disabled:text-disabled',
      'transition-[background-color,scale] ease-out-quad not-disabled:active:scale-98',
    ],
  ],
  variants: {
    variant: {
      primary: ['bg-brand text-fg-2', 'not-disabled:hover:bg-brand/90'],
      secondary: [
        'inset-ring-border bg-surface-2 text-fg-1',
        'not-disabled:hover:bg-fill-2',
      ],
      ghost: [
        'bg-transparent text-brand transition-[color_box-shadow]',
        'not-disabled:hover:inset-ring-2 not-disabled:hover:text-brand/90',
      ],
      destructive: [
        'inset-ring-transparent bg-danger text-fg-2',
        'not-disabled:hover:bg-danger/90',
      ],
    },
    size: {
      sm: 'h-9 gap-1 px-3 text-sm/5.5 [&_svg]:size-5',
      md: 'h-11 gap-2 px-4 text-base [&_svg]:size-6',
      lg: 'h-11 px-4 text-lg/7 [&_svg]:size-7',
      xl: 'h-14 px-6 text-xl/8 [&_svg]:size-8',
    },
    iconOnly: {
      true: 'aspect-square px-0',
    },
  },
  compoundVariants: [
    {
      variant: ['primary', 'ghost'],
      className: 'inset-ring-brand',
    },
    {
      size: ['lg', 'xl'],
      className: 'gap-3',
    },
  ],
  defaultVariants: {
    variant: 'primary',
    size: 'md',
    iconOnly: false,
  },
})

function Button({ className, variant, size, iconOnly, ...props }: ButtonProps) {
  return (
    <ark.button
      {...props}
      className={cn(
        buttonVariants({
          className,
          variant,
          size,
          iconOnly,
        }),
      )}
      data-scope="button"
      data-focusable
    />
  )
}

export { Button, buttonVariants }
export type { ButtonProps }
