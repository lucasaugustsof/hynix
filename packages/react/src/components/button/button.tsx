import { cn } from '@/utilities/cn'

export function Button({ children }: { children: React.ReactNode }) {
  return (
    <button type="button" className={cn('bg-blue-500 text-white px-4 py-2 rounded')}>
      {children}
    </button>
  )
}
