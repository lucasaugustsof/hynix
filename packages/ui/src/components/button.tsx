export function Button() {
  return (
    <button className="" type="button">
      Button
    </button>
  )
}
