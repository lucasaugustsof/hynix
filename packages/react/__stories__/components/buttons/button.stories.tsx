import type { Meta, StoryObj } from '@storybook/react'
import { expect, fn, userEvent, within } from '@storybook/test'

import { Button, type ButtonProps } from '@/registry/components/buttons/button'

import {
  RiArrowLeftLine,
  RiArrowRightLine,
  RiGithubFill,
} from '@remixicon/react'

const meta: Meta<ButtonProps> = {
  title: 'components/Buttons/Button',
  component: Button,
  args: {
    variant: 'primary',
    size: 'md',
    children: 'Button',
    disabled: false,
    iconOnly: false,
    onClick: fn(),
  },
  argTypes: {
    variant: {
      control: 'inline-radio',
      options: ['primary', 'secondary', 'ghost', 'destructive'],
      description: 'The visual style variant of the button',
    },
    size: {
      control: 'inline-radio',
      options: ['sm', 'md', 'lg', 'xl'],
      description: 'The size of the button',
    },
    disabled: {
      description: 'Whether the button is disabled and non-interactive',
    },
    iconOnly: {
      control: false,
      description:
        'Whether the button contains only an icon (requires aria-label)',
    },
    asChild: {
      control: false,
      description:
        'Whether to merge props onto a child element instead of creating a new button element',
    },
  },
  parameters: {
    layout: 'centered',
  },
  play: async ({ canvasElement, args }) => {
    const canvas = within(canvasElement)

    const sut = canvas.getByRole('button')

    await userEvent.click(sut)
    await userEvent.click(canvasElement)

    await expect(args.onClick).toHaveBeenCalled()
  },
}

export default meta

export const Basic: StoryObj<ButtonProps> = {}

export const WithPrefix: StoryObj<ButtonProps> = {
  render({ ...args }) {
    return (
      <Button {...args}>
        <RiArrowLeftLine />
        {args.children}
      </Button>
    )
  },
}

export const WithSuffix: StoryObj<ButtonProps> = {
  render({ ...args }) {
    return (
      <Button {...args}>
        {args.children}
        <RiArrowRightLine />
      </Button>
    )
  },
}

export const WithPrefixAndSuffix: StoryObj<ButtonProps> = {
  render({ ...args }) {
    return (
      <Button {...args}>
        <RiArrowLeftLine />
        {args.children}
        <RiArrowRightLine />
      </Button>
    )
  },
}

export const IconOnly: StoryObj<ButtonProps> = {
  args: {
    iconOnly: true,
    'aria-label': 'Go to GitHub',
  },
  argTypes: {
    children: {
      control: false,
    },
  },
  render({ ...args }) {
    return (
      <Button {...args}>
        <RiGithubFill />
      </Button>
    )
  },
}

export const LinkBehavior: StoryObj<ButtonProps> = {
  tags: ['skip-test'],
  args: {
    iconOnly: true,
    'aria-label': 'Go to GitHub',
    asChild: true,
  },
  argTypes: {
    children: {
      control: false,
    },
    disabled: {
      control: false,
    },
  },
  render({ ...args }) {
    return (
      <Button {...args} data-testid="github-link">
        <a href="https://github.com/lucasaugustsof/hynix">
          <RiGithubFill />
        </a>
      </Button>
    )
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement)

    const sut = canvas.getByTestId('github-link')

    await expect(sut.tagName.toLocaleLowerCase()).toEqual('a')
  },
}

export const Disabled: StoryObj<ButtonProps> = {
  args: {
    disabled: true,
    onClick: fn(),
  },
  play: async ({ canvasElement, args }) => {
    const canvas = within(canvasElement)

    const sut = canvas.getByRole('button')

    await userEvent.click(sut)
    await userEvent.click(canvasElement)

    await expect(args.onClick).not.toHaveBeenCalled()
  },
}
