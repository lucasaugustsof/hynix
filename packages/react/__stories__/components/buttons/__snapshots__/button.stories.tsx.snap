// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Buttons/Button Basic smoke-test 1`] = `
<button class="isolate inline-flex items-center justify-center whitespace-nowrap rounded-3xl font-semibold transition-colors ease-in-out focus-visible:outline-3 focus-visible:outline-brand-selected [&amp;_svg]:pointer-events-none [&amp;_svg]:shrink-0 inset-ring-brand bg-brand text-fg-2 enabled:hover:bg-brand-hover h-11 min-h-11 gap-2 px-4 text-base only:[&amp;_svg]:size-6 inset-ring-1 cursor-pointer"
        tabindex="0"
>
  Button
</button>
`;

exports[`components/Buttons/Button Disabled play-test 1`] = `
<button class="isolate inline-flex items-center justify-center whitespace-nowrap rounded-3xl font-semibold transition-colors ease-in-out focus-visible:outline-3 focus-visible:outline-brand-selected [&amp;_svg]:pointer-events-none [&amp;_svg]:shrink-0 inset-ring-brand enabled:hover:bg-brand-hover h-11 min-h-11 gap-2 px-4 text-base only:[&amp;_svg]:size-6 cursor-not-allowed bg-fill-1 text-disabled [&amp;_svg]:text-disabled"
        disabled
        tabindex="-1"
>
  Button
</button>
`;

exports[`components/Buttons/Button IconOnly smoke-test 1`] = `
<button aria-label="Go to GitHub"
        class="isolate inline-flex items-center justify-center whitespace-nowrap rounded-3xl font-semibold transition-colors ease-in-out focus-visible:outline-3 focus-visible:outline-brand-selected [&amp;_svg]:pointer-events-none [&amp;_svg]:shrink-0 inset-ring-brand bg-brand text-fg-2 enabled:hover:bg-brand-hover h-11 min-h-11 gap-2 text-base only:[&amp;_svg]:size-6 inset-ring-1 cursor-pointer aspect-square px-0"
        tabindex="0"
>
  <svg viewbox="0 0 24 24"
       xmlns="http://www.w3.org/2000/svg"
       width="24"
       height="24"
       fill="currentColor"
       class="remixicon "
  >
    <path d="M12.001 2C6.47598 2 2.00098 6.475 2.00098 12C2.00098 16.425 4.86348 20.1625 8.83848 21.4875C9.33848 21.575 9.52598 21.275 9.52598 21.0125C9.52598 20.775 9.51348 19.9875 9.51348 19.15C7.00098 19.6125 6.35098 18.5375 6.15098 17.975C6.03848 17.6875 5.55098 16.8 5.12598 16.5625C4.77598 16.375 4.27598 15.9125 5.11348 15.9C5.90098 15.8875 6.46348 16.625 6.65098 16.925C7.55098 18.4375 8.98848 18.0125 9.56348 17.75C9.65098 17.1 9.91348 16.6625 10.201 16.4125C7.97598 16.1625 5.65098 15.3 5.65098 11.475C5.65098 10.3875 6.03848 9.4875 6.67598 8.7875C6.57598 8.5375 6.22598 7.5125 6.77598 6.1375C6.77598 6.1375 7.61348 5.875 9.52598 7.1625C10.326 6.9375 11.176 6.825 12.026 6.825C12.876 6.825 13.726 6.9375 14.526 7.1625C16.4385 5.8625 17.276 6.1375 17.276 6.1375C17.826 7.5125 17.476 8.5375 17.376 8.7875C18.0135 9.4875 18.401 10.375 18.401 11.475C18.401 15.3125 16.0635 16.1625 13.8385 16.4125C14.201 16.725 14.5135 17.325 14.5135 18.2625C14.5135 19.6 14.501 20.675 14.501 21.0125C14.501 21.275 14.6885 21.5875 15.1885 21.4875C19.259 20.1133 21.9999 16.2963 22.001 12C22.001 6.475 17.526 2 12.001 2Z">
    </path>
  </svg>
</button>
`;

exports[`components/Buttons/Button WithPrefix smoke-test 1`] = `
<button class="isolate inline-flex items-center justify-center whitespace-nowrap rounded-3xl font-semibold transition-colors ease-in-out focus-visible:outline-3 focus-visible:outline-brand-selected [&amp;_svg]:pointer-events-none [&amp;_svg]:shrink-0 inset-ring-brand bg-brand text-fg-2 enabled:hover:bg-brand-hover h-11 min-h-11 gap-2 px-4 text-base only:[&amp;_svg]:size-6 inset-ring-1 cursor-pointer"
        tabindex="0"
>
  <svg viewbox="0 0 24 24"
       xmlns="http://www.w3.org/2000/svg"
       width="24"
       height="24"
       fill="currentColor"
       class="remixicon "
  >
    <path d="M7.82843 10.9999H20V12.9999H7.82843L13.1924 18.3638L11.7782 19.778L4 11.9999L11.7782 4.22168L13.1924 5.63589L7.82843 10.9999Z">
    </path>
  </svg>
  Button
</button>
`;

exports[`components/Buttons/Button WithPrefixAndSuffix smoke-test 1`] = `
<button class="isolate inline-flex items-center justify-center whitespace-nowrap rounded-3xl font-semibold transition-colors ease-in-out focus-visible:outline-3 focus-visible:outline-brand-selected [&amp;_svg]:pointer-events-none [&amp;_svg]:shrink-0 inset-ring-brand bg-brand text-fg-2 enabled:hover:bg-brand-hover h-11 min-h-11 gap-2 px-4 text-base only:[&amp;_svg]:size-6 inset-ring-1 cursor-pointer"
        tabindex="0"
>
  <svg viewbox="0 0 24 24"
       xmlns="http://www.w3.org/2000/svg"
       width="24"
       height="24"
       fill="currentColor"
       class="remixicon "
  >
    <path d="M7.82843 10.9999H20V12.9999H7.82843L13.1924 18.3638L11.7782 19.778L4 11.9999L11.7782 4.22168L13.1924 5.63589L7.82843 10.9999Z">
    </path>
  </svg>
  Button
  <svg viewbox="0 0 24 24"
       xmlns="http://www.w3.org/2000/svg"
       width="24"
       height="24"
       fill="currentColor"
       class="remixicon "
  >
    <path d="M16.1716 10.9999L10.8076 5.63589L12.2218 4.22168L20 11.9999L12.2218 19.778L10.8076 18.3638L16.1716 12.9999H4V10.9999H16.1716Z">
    </path>
  </svg>
</button>
`;

exports[`components/Buttons/Button WithSuffix smoke-test 1`] = `
<button class="isolate inline-flex items-center justify-center whitespace-nowrap rounded-3xl font-semibold transition-colors ease-in-out focus-visible:outline-3 focus-visible:outline-brand-selected [&amp;_svg]:pointer-events-none [&amp;_svg]:shrink-0 inset-ring-brand bg-brand text-fg-2 enabled:hover:bg-brand-hover h-11 min-h-11 gap-2 px-4 text-base only:[&amp;_svg]:size-6 inset-ring-1 cursor-pointer"
        tabindex="0"
>
  Button
  <svg viewbox="0 0 24 24"
       xmlns="http://www.w3.org/2000/svg"
       width="24"
       height="24"
       fill="currentColor"
       class="remixicon "
  >
    <path d="M16.1716 10.9999L10.8076 5.63589L12.2218 4.22168L20 11.9999L12.2218 19.778L10.8076 18.3638L16.1716 12.9999H4V10.9999H16.1716Z">
    </path>
  </svg>
</button>
`;
